src/middlewared/middlewared/plugins/nvmet/global.py

```diff
@@ -0,0 +1,275 @@
+        if await self.ana_enabled():
+            sessions.extend(await self.middleware.call('failover.call_remote',
+                                                       'nvmet.global.local_sessions',
+                                                       [subsys_id]))
+
+        return filter_list(sessions, filters, options)
+
+    def __parse_session_dir(self, path: pathlib.Path, port_index_to_id: dict):
+        if path.name.startswith('ctrl'):
+            result = {}
+            result['ctrl'] = int(path.name[4:])
+            result['hostnqn'] = pathlib.Path(path, 'hostnqn').read_text().strip()
+            result['host_traddr'] = pathlib.Path(path, 'host_traddr').read_text().strip()
+            port_index = int(pathlib.Path(path, 'port').read_text().strip())
+            result['port_id'] = port_index_to_id[port_index]
+            return result
+
+    @private
+    def local_sessions(self, subsys_id=None):
+        sessions = []
+        global_info = self.middleware.call_sync('nvmet.global.config')
+        subsystems = self.middleware.call_sync('nvmet.subsys.query')
+
+        port_index_to_id = {port['index']: port['id'] for port in self.middleware.call_sync('nvmet.port.query')}
+
+        if subsys_id is None:
+            basenqn = global_info['basenqn']
+            subsys_name_to_subsys_id = {f'{basenqn}:{subsys["name"]}': subsys['id'] for subsys in subsystems}
+            for subsys in pathlib.Path(NVMET_DEBUG_DIR).iterdir():
+                if subsys_id := subsys_name_to_subsys_id.get(subsys.name):
+                    for ctrl in subsys.iterdir():
+                        if session := self.__parse_session_dir(ctrl, port_index_to_id):
+                            session['subsys_id'] = subsys_id
+                            sessions.append(session)
+        else:
+            for subsys in subsystems:
+                if subsys['id'] == subsys_id:
+                    subnqn = f'{global_info["basenqn"]}:{subsys["name"]}'
+                    path = pathlib.Path(NVMET_DEBUG_DIR, subnqn)
+                    if path.is_dir():
+                        for ctrl in path.iterdir():
+                            if session := self.__parse_session_dir(ctrl, port_index_to_id):
+                                session['subsys_id'] = subsys_id
+                                sessions.append(session)
+
+        return sessions
+
+    @private
+    async def load_kernel_modules(self):
+        if (await self.config())['kernel']:
+            do_load = any([
+                await self.middleware.call('failover.is_single_master_node'),
+                await self.middleware.call('nvmet.global.ana_active')
+            ])
+            if do_load:
+                modules = await self.__kernel_modules()
+                await self.middleware.run_in_thread(load_modules, modules)
+
+    @private
+    async def unload_kernel_modules(self):
+        if modules := await self.__kernel_modules():
```

Comment:
Will it skip unloading some modules if the configuration changed before calling start and stop?

Reply:
Yes

I was thinking about protecting the `kernel` setting so that it could only be changed when things are stopped.  (Currently it can't be changed **at all**.)

However, you're correct that `rdma_enabled` setting shouldn't factor in, and we should always unload it if loaded.  Thanks.